File: src/contraction_controller_px4/contraction_controller_px4/run_node.py
"""Entry point for the contraction controller ROS2 node."""

import argparse
import os
import traceback

import rclpy

from quad_platforms import PlatformType  # type: ignore[import]
from .trajectories import TrajectoryType
from .ros2px4_node import ContractionOffboardControl

from Logger import Logger  # type: ignore[import]
from Logger.shutdown_helpers import install_shutdown_logging  # type: ignore[import]


def create_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Contraction Neural-Network Offboard Controller for Quadrotor",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  ros2 run contraction_controller_px4 run_node --platform sim --trajectory hover --hover-mode 1
  ros2 run contraction_controller_px4 run_node --platform sim --trajectory figure_eight --log
  ros2 run contraction_controller_px4 run_node --platform sim --trajectory trefoil --log --log-file my_run
  ros2 run contraction_controller_px4 run_node --platform hw  --trajectory spiral
        """,
    )

    parser.add_argument(
        "--platform",
        type=PlatformType,
        choices=list(PlatformType),
        required=True,
        help="Platform: " + ", ".join(e.value for e in PlatformType),
    )
    parser.add_argument(
        "--trajectory",
        type=TrajectoryType,
        choices=list(TrajectoryType),
        required=True,
        help="Trajectory: " + ", ".join(e.value for e in TrajectoryType),
    )
    parser.add_argument(
        "--hover-mode",
        type=int,
        choices=range(1, 9),
        help="Hover sub-mode (required when --trajectory=hover).",
    )
    parser.add_argument(
        "--controller-dir",
        type=str,
        default=None,
        help="Path to the Controller directory containing arch.txt and model.eqx. "
             "Defaults to src/controller_params relative to workspace root.",
    )
    parser.add_argument(
        "--flight-period",
        type=float,
        default=None,
        help="Flight duration in seconds (default: 30s sim / 60s hw).",
    )
    parser.add_argument(
        "--log",
        action="store_true",
        help="Enable CSV data logging. Written to src/data_analysis/log_files/ on shutdown.",
    )
    parser.add_argument(
        "--log-file",
        type=str,
        default=None,
        help="Log filename stem (without .csv). Auto-generated from config if omitted.",
    )

    return parser


def validate_args(args, parser: argparse.ArgumentParser) -> None:
    if args.trajectory == TrajectoryType.HOVER:
        if args.hover_mode is None:
            parser.error("--hover-mode is required when --trajectory=hover")
    else:
        if args.hover_mode is not None:
            parser.error("--hover-mode is only valid when --trajectory=hover")
    if args.log_file is not None and not args.log:
        parser.error("--log-file requires --log")


def _auto_log_filename(args) -> str:
    parts = [args.platform.value, "contraction", args.trajectory.value]
    if args.trajectory == TrajectoryType.HOVER and args.hover_mode is not None:
        parts.append(f"mode{args.hover_mode}")
    return "_".join(parts) + ".csv"


def main():
    parser = create_parser()
    args = parser.parse_args()
    validate_args(args, parser)

    logging_enabled = args.log
    if logging_enabled:
        log_file = args.log_file if args.log_file else _auto_log_filename(args)
        if not log_file.lower().endswith(".csv"):
            log_file += ".csv"
    else:
        log_file = None

    print("\n" + "=" * 60)
    print("Contraction Neural-Network Offboard Controller")
    print("=" * 60)
    print(f"Platform:      {args.platform.value.upper()}")
    print(f"Trajectory:    {args.trajectory.value.upper()}")
    print(f"Hover Mode:    {args.hover_mode if args.hover_mode is not None else 'N/A'}")
    print(f"Flight Period: {args.flight_period or ('30s' if args.platform == PlatformType.SIM else '60s')}")
    print(f"Controller:    {args.controller_dir or '(default: src/controller_params)'}")
    print(f"Data Logging:  {'Enabled → ' + log_file if logging_enabled else 'Disabled'}")
    print("=" * 60 + "\n")

    base_path = os.path.dirname(os.path.abspath(__file__))

    rclpy.init(args=None)
    node = ContractionOffboardControl(
        platform_type=args.platform,
        trajectory=args.trajectory,
        hover_mode=args.hover_mode,
        controller_dir=args.controller_dir,
        flight_period_=args.flight_period,
        logging_enabled=logging_enabled,
    )

    logger = Logger(log_file, base_path) if logging_enabled else None

    if logger is not None:
        install_shutdown_logging(logger, node)

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        print("\nKeyboard interrupt.")
    except Exception as e:
        print(f"\nError: {e}")
        traceback.print_exc()
    finally:
        if logger is not None and logging_enabled:
            print("Saving log data...")
            logger.log(node)
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()
        print("Node shut down.")


if __name__ == "__main__":
    main()

File: src/nr_diff_flat_px4_cpp/src/run_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <iostream>
#include <string>
#include <algorithm>
#include <optional>
#include <filesystem>

#include <ros2_logger_cpp/log_path.hpp>

#include "quad_platforms_cpp/platform_config.hpp"
#include "quad_trajectories_cpp/types.hpp"
#include "nr_diff_flat_px4_cpp/offboard_control_node.hpp"

namespace qp = quad_platforms_cpp;
namespace qt = quad_trajectories_cpp;

struct Args {
    qp::PlatformType platform;
    qt::TrajectoryType trajectory;
    std::optional<int> hover_mode;
    bool log = false;
    std::string log_file;
    bool double_speed = false;
    bool short_variant = false;
    bool spin = false;
    std::optional<double> flight_period;
    bool feedforward = false;
    std::string nr_profile = "baseline";
};

static std::string generate_log_filename(const Args& args) {
    std::string name = qp::platform_type_to_string(args.platform);
    name += "_nr_diff_flat_";
    name += qt::trajectory_type_to_string(args.trajectory);
    if (args.feedforward) name += "_ff";
    if (args.nr_profile != "baseline") name += "_" + args.nr_profile;
    name += args.double_speed ? "_2x" : "_1x";
    if (args.short_variant) name += "_short";
    if (args.spin) name += "_spin";
    name += "_cpp";
    return name;
}

static void print_usage() {
    std::cerr <<
        "\nUsage: ros2 run nr_diff_flat_px4_cpp run_node [OPTIONS]\n"
        "\n"
        "Newton-Raphson Diff-Flat Offboard Control for Quadrotor\n"
        "\n"
        "Required:\n"
        "  --platform {sim,hw}          Platform type\n"
        "  --trajectory {hover,yaw_only,circle_horz,circle_vert,\n"
        "                fig8_horz,fig8_vert,triangle,sawtooth,\n"
        "                helix,fig8_contraction}\n"
        "                               Trajectory to execute\n"
        "\n"
        "Optional:\n"
        "  --hover-mode 1-8             Required when --trajectory=hover (hw: 1-4 only)\n"
        "  --log                        Enable data logging (auto-generates filename)\n"
        "  --log-file NAME              Custom log filename (requires --log)\n"
        "  --double-speed               Use 2x speed for trajectories\n"
        "  --short                      Use short variant for fig8_vert trajectory\n"
        "  --spin                       Enable spin for circle_horz and helix\n"
        "  --flight-period SECONDS      Override default flight duration (sim: 30s, hw: 60s)\n"
        "  --ff                         Enable feedforward (only valid with --trajectory=fig8_contraction)\n"
        "  --nr-profile {baseline,workshop}\n"
        "                               Newton-Raphson profile to run\n"
        "\n"
        "Examples:\n"
        "  # Auto-generated log filename:\n"
        "  ros2 run nr_diff_flat_px4_cpp run_node --platform sim --trajectory helix --double-speed --spin --log\n"
        "  # -> logs to: src/data_analysis/log_files/nr_diff_flat_px4_cpp/sim_nr_diff_flat_helix_2x_spin.csv\n"
        "\n"
        "  # Custom log filename:\n"
        "  ros2 run nr_diff_flat_px4_cpp run_node --platform sim --trajectory helix --log --log-file my_log\n"
        "  # -> logs to: src/data_analysis/log_files/nr_diff_flat_px4_cpp/my_log.csv\n"
        "\n"
        "  # Feedforward (fig8_contraction only):\n"
        "  ros2 run nr_diff_flat_px4_cpp run_node --platform sim --trajectory fig8_contraction --ff --log\n"
        "  # -> logs to: src/data_analysis/log_files/nr_diff_flat_px4_cpp/sim_nr_diff_flat_fig8_contraction_ff_1x.csv\n"
        "\n";
}

static Args parse_args(int argc, char* argv[]) {
    Args args;
    bool has_platform = false, has_trajectory = false;

    for (int i = 1; i < argc; ++i) {
        std::string arg = argv[i];

        if (arg == "--platform" && i + 1 < argc) {
            args.platform = qp::platform_type_from_string(argv[++i]);
            has_platform = true;
        } else if (arg == "--trajectory" && i + 1 < argc) {
            args.trajectory = qt::trajectory_type_from_string(argv[++i]);
            has_trajectory = true;
        } else if (arg == "--hover-mode" && i + 1 < argc) {
            args.hover_mode = std::stoi(argv[++i]);
        } else if (arg == "--log") {
            args.log = true;
        } else if (arg == "--log-file" && i + 1 < argc) {
            args.log_file = argv[++i];
        } else if (arg == "--double-speed") {
            args.double_speed = true;
        } else if (arg == "--short") {
            args.short_variant = true;
        } else if (arg == "--spin") {
            args.spin = true;
        } else if (arg == "--flight-period" && i + 1 < argc) {
            args.flight_period = std::stod(argv[++i]);
        } else if (arg == "--ff") {
            args.feedforward = true;
        } else if (arg == "--nr-profile" && i + 1 < argc) {
            args.nr_profile = argv[++i];
        } else {
            print_usage();
            throw std::runtime_error("Unknown argument: " + arg);
        }
    }

    if (!has_platform || !has_trajectory) {
        print_usage();
        throw std::runtime_error("--platform and --trajectory are required");
    }

    // Validation
    if (args.trajectory == qt::TrajectoryType::HOVER) {
        if (!args.hover_mode.has_value()) {
            throw std::runtime_error("--hover-mode is required when --trajectory=hover");
        }
        if (args.platform == qp::PlatformType::HARDWARE && args.hover_mode.value() > 4) {
            throw std::runtime_error("--hover-mode must be 1-4 for --platform=hw");
        }
    } else {
        if (args.hover_mode.has_value()) {
            throw std::runtime_error("--hover-mode is only valid when --trajectory=hover");
        }
    }

    if (!args.log_file.empty() && !args.log) {
        throw std::runtime_error("--log-file requires --log to be enabled");
    }

    if (args.feedforward && args.trajectory != qt::TrajectoryType::FIG8_CONTRACTION) {
        throw std::runtime_error("--ff is only valid with --trajectory=fig8_contraction");
    }
    if (args.nr_profile != "baseline" && args.nr_profile != "workshop") {
        throw std::runtime_error("--nr-profile must be baseline or workshop");
    }

    return args;
}

int main(int argc, char* argv[]) {
    Args args;
    try {
        args = parse_args(argc, argv);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }

    // Determine log file path using ros2_logger_cpp path resolution
    std::string log_file;
    if (args.log) {
        std::string fname = args.log_file.empty() ? generate_log_filename(args) : args.log_file;
        std::filesystem::path exe_path = std::filesystem::canonical(argv[0]);
        auto paths = ros2_logger_cpp::resolve_log_path(fname, exe_path);
        log_file = paths.log_file.string();
        ros2_logger_cpp::copy_analysis_helpers(paths.data_analysis);
    }

    // Print configuration
    std::string sep(60, '=');
    std::cout << "\n" << sep << std::endl;
    std::cout << "Newton-Raphson Diff-Flat Offboard Control Configuration (C++)" << std::endl;
    std::cout << sep << std::endl;
    std::cout << "Platform:      " << qp::platform_type_to_string(args.platform) << std::endl;
    std::cout << "Controller:    STANDARD" << std::endl;
    std::cout << "Trajectory:    " << qt::trajectory_type_to_string(args.trajectory) << std::endl;
    std::cout << "Hover Mode:    " << (args.hover_mode ? std::to_string(*args.hover_mode) : "N/A") << std::endl;
    std::cout << "Speed:         " << (args.double_speed ? "Double (2x)" : "Regular (1x)") << std::endl;
    std::cout << "Short:         " << (args.short_variant ? "Enabled" : "Disabled") << std::endl;
    double fp = args.flight_period.value_or(
        args.platform == qp::PlatformType::HARDWARE ? 60.0 : 30.0);
    std::cout << "Flight Period: " << fp << " seconds" << std::endl;
    std::cout << "Spin:          " << (args.spin ? "Enabled" : "Disabled") << std::endl;
    std::cout << "Feedforward:   " << (args.feedforward ? "Enabled (fig8_contraction)" : "Disabled") << std::endl;
    std::cout << "NR Profile:    " << args.nr_profile << std::endl;
    std::cout << "Data Logging:  " << (args.log ? "Enabled" : "Disabled") << std::endl;
    if (args.log) {
        std::cout << "Log File:      " << log_file << std::endl;
    }
    std::cout << sep << "\n" << std::endl;

    rclcpp::init(argc, argv);

    auto node = std::make_shared<nr_diff_flat_px4_cpp::OffboardControlNode>(
        args.platform, args.trajectory, args.hover_mode,
        args.double_speed, args.short_variant, args.spin,
        args.log, log_file, args.flight_period, args.feedforward, args.nr_profile);

    try {
        rclcpp::spin(node);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    // Log is saved automatically in OffboardControlNode's destructor when node goes out of scope.

    rclcpp::shutdown();
    std::cout << "\nNode shut down." << std::endl;
    return 0;
}
